spark/src/main/scala/io/substrait/spark/expression/ToSubstraitLiteral.scala

```diff
@@ -34,6 +38,34 @@ class ToSubstraitLiteral {
         scale: Int): SExpression.Literal =
       decimal(false, d.toJavaBigDecimal, precision, scale)
 
+    private def sparkArray2Substrait(
+        arrayData: ArrayData,
+        elementType: DataType,
+        containsNull: Boolean): SExpression.Literal = {
+      val elements = arrayData.array.map(any => apply(Literal(any, elementType)))
+      if (elements.isEmpty) {
+        return emptyList(false, ToSubstraitType.convert(elementType, nullable = containsNull).get)
+      }
+      list(false, JavaConverters.asJavaIterable(elements)) // TODO: handle containsNull
```

Comment:
Handling `containsNull` here is non-trivial, since Substrait doesn't have such a concept. The nullability of the non-empty list is taken from the nullability of the element type, which is gotten from the first element. So we'll need to encode `containsNull` into the `elementType`. It can be done, but requires some more work (as there's no "withNullable" method or anything) and this PR is already big enough, so I'll do it as a followup.

Reply:
I'm not entirely sure what you mean here.

> The nullability of the non-empty list is taken from the nullability of the element type, which is gotten from the first element.

Is this when going from Substrait to Spark or from Spark to Substrait?

Reply:
Substrait itself doesn't contain the "containsNull" arg, so the Substrait -> Spark conversion has to infer it, and the Spark -> Substrait conversion doesn't provide the correct information to infer it based on.

Easiest to see from the commented-out test:
Starting with Spark: `value: [1,null,3] type: ArrayType(IntegerType,true)`
Gets converted into Substrait literal: `ListLiteral{nullable=false, values=[I32Literal{nullable=false, value=1}, NullLiteral{nullable=false, type=I32{nullable=true}}, I32Literal{nullable=false, value=3}]}`

Note that there is no "containsNull" anywhere. So the way we infer it when converting back into Substrait is to look at the first element and it's `nullable` arg. Which in today's world is always `nullabe=false`, leading to:

Spark: `value: [1,null,3] type: ArrayType(IntegerType,false)`

(Funnily enough that's the case even if the first element _is_ a null.. but fixing that in itself wouldn't be enough, since the null might be elsewhere in the list or even in a different row, but we should have the "type" of the literal match across rows.)

I think fixing this requires changing the `ToSubstraitLiteral.convert/convertWithValue` so that the nullability they set can be overridden in this case.